### PR TITLE
 Generic FlightTableFactory with a default FlightSqlDriver

### LINKED ISSUE #7731: Arrow Flight SQL / ADBC datasource
### Is your feature request related to a problem or challenge?

I want to fetch data from multiple remote databases.

### Describe the solution you'd like

Since DataFusion is part of the Arrow ecosystem it seems to make sense to have an Arrow Flight SQL datasource.

### Describe alternatives you've considered

[splitgraph/seafowl's datafusion_remote_tables](https://github.com/splitgraph/seafowl/tree/main/datafusion_remote_tables). However, it seems sensible to have a basic Arrow Flight SQL datasource out of the box.

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by ccciudatu**
## Which issue does this PR close?

Closes #7731

## Rationale for this change

Create external tables backed by Arrow Flight(SQL) services.

## What changes are included in this PR?
- A generic `FlightTableFactory` that can integrate any Arrow Flight RPC service as a `TableProviderFactory`, expecting a `FlightDriver` implementation to handle the `GetFlightInfo` flow.
- Default `FlightSqlDriver` implementation registered as `FLIGHT_SQL`.

## Are these changes tested?
yes

### CODE REVIEW

**Comment by alamb** [comment]
Thank you @ccciudatu 

I have been thinking recently that we need to figure out what to do with `datafusion-cli` -- it seems to currently be serving two roles:
1. Debug tool for playing around with supported sql, making reproducers
2. An actual CLI tool with pre-built integrations (I am thinking specifically of https://github.com/apache/datafusion/pull/10792 from @xinlifoobar  here)

There is a tension for adding integrations into datafusion-cli as it makes the overall project that much more complicated

@stuartcarnie  and I were am thinking of writing up a proposal to make a  `dfdb`, or `dfcli` binary where we could start adding all these pre-built integrations. I think that would be quite compelling

I plan to try to write this up / maybe try and make an example repo or something

**Comment by ccciudatu** [comment]
@alamb The PR title is misleading; this is *not* meant to be a `datafusion-cli` patch. It only got renamed because I added a follow-up commit to make the new `FLIGHT_SQL` table "storage" type work with `datafusion-cli` (i.e. register the new config namespace).
However, I'm glad this triggered such an interesting high-level discussion about revisiting the cli package. :)

I will squash the whole thing in a single commit and restore the original title .
Also, I just realized that I need to add proto serialization for the new FlightExec physical plan, so I'll convert this the PR to a draft for now.

**Comment by ccciudatu** [comment]
I added the proto marshalling and marked the PR "ready for review", once again.

File: datafusion/proto/proto/datafusion.proto [hunk 1]
```diff
@@ -1199,3 +1200,28 @@ message PartitionStats {
...
+message PlanPropertiesNode {
+  datafusion_common.Schema schema = 1;
+  repeated PhysicalSortExprNodeCollection output_ordering = 2;
+  Partitioning partitioning = 3;
+  ExecutionMode execution_mode = 4;
+}
+
+enum ExecutionMode {
+  Bounded = 0;
+  Unbonded = 1;
```
Comment (ccciudatu) [inline datafusion/proto/proto/datafusion.proto:1224]: Just noticed this typo :(

File: datafusion/proto/src/physical_plan/mod.rs [hunk 2]
```diff
@@ -1081,6 +1091,67 @@ impl AsExecutionPlan for protobuf::PhysicalPlanNode {
...
+                    _ => unreachable!("Unexpected execution mode"),
+                };
+                let plan_properties =
+                    PlanProperties::new(eq_props, partitioning, execution_mode);
+
+                Ok(Arc::new(FlightExec::restore(
+                    partitions,
+                    plan_properties,
+                    &flight_scan.grpc_headers,
+                )))
```
Comment (ccciudatu) [inline datafusion/proto/src/physical_plan/mod.rs:1153]: The feature flag above will fail the build when the "flight" feature is disabled. I'll add an "else" to let it compile and also fix it for the "parquet" feature flag.
However, there are other places where disabling `parquet` will fail the build, but I don't want to expand the scope of this PR.

Comment (ccciudatu) [reply datafusion/proto/src/physical_plan/mod.rs:1161]: fixed

**Comment by alamb** [comment]
> @alamb The `[datafusion-cli]` commit was just a follow-up to make the new `FLIGHT_SQL` table "storage" type available in `datafusion-cli` (i.e. register the new config namespace). I'm sorry if that was misleading. The patch is supposed to add support for Arrow Flight (SQL) data sources.

FWI I wrote up my thoughts here; https://github.com/apache/datafusion/issues/11979

I hope to review this PR later today

One thing that might be useful to ask is "does this code need to be in the datafusion repo, or could it be put in another repo -- like for example https://github.com/datafusion-contrib/datafusion-table-providers 🤔 

We could make you your own repo if you wanted

**Comment by ccciudatu** [comment]
@alamb Got it, makes sense.
My initial thought was that having this in the main repo would enable a better overall integration (i.e. datafusion-cli support, transparent proto serialization, etc.) but I guess having this in a contrib project should not limit these options. I'll have a better look at the project you linked to see if that's a better fit, but I'll go with whatever you decide if you come to a conclusion.

Separately, I promise I'll stop force-pushing updates now that I know it's being reviewed. I may push some fixups though...  :)
The last change was to account for server implementations that do not allow reusing the same FlightInfo object, so now GetFlightInfo is called before every query against a "flight table", rather than only once on table creation.
Also, I found a weird/buggy behaviour related to some schema mismatch, i.e. columns seem to get reordered in certain scenarios. I will try to reproduce that in a test and follow up with the fix or at least the failing test to reproduce it.

**Comment by ccciudatu** [comment]
I found my embarrassing mistake and added a fixup commit.

File: datafusion/core/src/datasource/flight/sql.rs [hunk 3]
```diff
@@ -0,0 +1,475 @@
...
+}
+
+/////////////////////////////////////////////////////////////////////////
+// Shameless copy/paste from arrow-flight FlightSqlServiceClient
+// This is only needed in order to access the bearer token received
+// during handshake, as the standard client does not expose this information.
+// The bearer token has to be passed to the clients that perform
+// the DoGet operation, since Dremio, Ballista and possibly others
+// expect the bearer token they produce with the handshake response
+// to be set on all subsequent requests, including DoGet.
```
Comment (ccciudatu) [inline datafusion/core/src/datasource/flight/sql.rs:93]: TODO: Remove this "stolen" client implementation and submit a PR to arrow-rs to make the token visible in the official client.

Comment (ccciudatu) [reply datafusion/core/src/datasource/flight/sql.rs:93]: https://github.com/apache/arrow-rs/issues/6253

Comment (ccciudatu) [reply datafusion/core/src/datasource/flight/sql.rs:93]: https://github.com/apache/arrow-rs/pull/6254

Comment (ccciudatu) [reply datafusion/core/src/datasource/flight/sql.rs:93]: @alamb on average, my PRs are of reasonable size. :)

Comment (alamb) [reply datafusion/core/src/datasource/flight/sql.rs:93]: I appreciate it @ccciudatu  -- I am sorry I haven't had a chance to comment more thoroughly on this PR yet...

Comment (ccciudatu) [reply datafusion/core/src/datasource/flight/sql.rs:93]: No worries, @alamb! The fact that you are even considering reviewing such a huge patch makes you one of the nicest people I know. 🙃

Comment (alamb) [reply datafusion/core/src/datasource/flight/sql.rs:93]: that is nice of you to say -- the whole point of doing this in open source is to get other people to contribute things and a key way to do that is to actually look at their proposals 😆

File: datafusion/core/src/datasource/flight/config.rs [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+//! Only meant for registering the `flight` namespace with `datafusion-cli`
+
+use datafusion_common::config::{ConfigEntry, ConfigExtension, ExtensionOptions};
+use std::any::Any;
+use std::collections::HashMap;
+
+/// Collects and reports back config entries. Only used to persuade `datafusion-cli`
+/// to accept the `flight.` prefix for `CREATE EXTERNAL TABLE` options.
+#[derive(Default, Debug, Clone)]
+pub struct FlightOptions {
```
Comment (alamb) [inline datafusion/core/src/datasource/flight/config.rs:27]: The configuration system is another thing I think that a focused standalone binary would be able to do better than DataFusion's key=value system

File: datafusion/core/src/datasource/flight/mod.rs [hunk 5]
```diff
@@ -0,0 +1,259 @@
...
+/// }
+///
+/// #[tokio::main]
+/// async fn main() -> datafusion_common::Result<()> {
+///     let ctx = SessionContext::new();
+///     ctx.state_ref().write().table_factories_mut()
+///         .insert("CUSTOM_FLIGHT".into(), Arc::new(FlightTableFactory::new(
+///             Arc::new(CustomFlightDriver::default())
+///         )));
+///     let _ = ctx.sql(r#"
```
Comment (alamb) [inline datafusion/core/src/datasource/flight/mod.rs:79]: this is really cool

File: datafusion/core/src/datasource/flight/mod.rs [hunk 6]
```diff
@@ -0,0 +1,259 @@
...
+    /// driver doesn't need (or care) to customize the execution plan.
+    pub(super) plan_properties: Arc<PlanProperties>,
+    /// The gRPC headers to use on the `DoGet` calls
+    pub(super) grpc_metadata: Arc<MetadataMap>,
+}
+
+impl FlightMetadata {
+    /// Provide custom [PlanProperties] to account for service specifics,
+    /// such as known partitioning scheme, unbounded execution mode etc.
+    pub fn new(info: FlightInfo, props: PlanProperties, grpc: MetadataMap) -> Self {
```
Comment (alamb) [inline datafusion/core/src/datasource/flight/mod.rs:170]: this is really cool -- we actually implemented some version of this custom for InfluxDB 3.0 for services that exchange data between themselves

File: datafusion/core/src/datasource/flight/sql.rs [hunk 7]
```diff
@@ -0,0 +1,475 @@
...
+use bytes::Bytes;
+use futures::{stream, TryStreamExt};
+use prost::Message;
+use tonic::metadata::{AsciiMetadataKey, MetadataMap};
+use tonic::transport::Channel;
+use tonic::IntoRequest;
+
+use crate::datasource::flight::{FlightDriver, FlightMetadata};
+
+/// Default Flight SQL driver. Requires a `flight.sql.query` to be passed as a table option.
```
Comment (alamb) [inline datafusion/core/src/datasource/flight/sql.rs:40]: Yeah, the configuration story for coinfiguring datasources like this is sorely lacking in datafusion-cli

It would be kind of cool to support something like

```sql
create crednetials for flightsql as ....

select * from 'flightsql://myhost/SELECT%20*%20FROM%20foo'
```

**Review by alamb — COMMENTED** [review]
First of all, thank you @ccciudatu for this contribution. It si

TLDR is I recommend we put this code into a separate repo and indeed have made one here:  https://github.com/datafusion-contrib/datafusion-table-provider-flight

The rationale is that since this code uses publically available APIs so there is no technical reason it must be in the DataFusion core crate that I can see

You raise an excellent point about a pre-integrated binary being  easier to use with `datafusion-cli`, and I agree such a pre-integration would be useful to users

However, given how much is already in the DataFusion core crates I think adding more features at this time is likely not a great idea. 

Instead, as I wrote down, I think it would be ideal if this provider (along with others) could be pre-integrated into a single binary via something like https://github.com/apache/datafusion/issues/11979

I left some suggestions along the way in case it is helpful to you

**Comment by ccciudatu** [comment]
Thank you, @alamb . I will move this to the new repo and close it from here.
It can't really be moved as such, as the proto stuff is cumbersome to replicate outside of this repo, so I think I'll switch to a serde-based codec for the new physical plan node.
The `datafusion-cli` integration was never a priority for me (I just thought it was the civilised way to add new table factories), so I have no problem living without it for now.

**Comment by alamb** [comment]
> It can't really be moved as such, as the proto stuff is cumbersome to replicate outside of this repo, so I think I'll switch to a serde-based codec for the new physical plan node.

Makes sense -- thank you for your understanding

**Comment by ccciudatu** [comment]
Moved to its own repo: https://github.com/datafusion-contrib/datafusion-table-provider-flight/pull/1
Thanks, @alamb !

**Comment by phillipleblanc** [comment]
@ccciudatu Just seeing this now - I'm one of the maintainers of https://github.com/datafusion-contrib/datafusion-table-providers. I think it would be useful to have a single crate of community maintained TableProviders and I think this would fit nicely in there, and I could make you a maintainer on that repo as well.

We've also done work to make the table providers in there play nicely with https://github.com/datafusion-contrib/datafusion-federation, which could also be good for this implementation. Also happy if you want to keep it in a separate crate/repo, we can link to it from README.md.

**Comment by ccciudatu** [comment]
You are right, @phillipleblanc, I enjoyed having my own repo for a while, but I soon started to feel lonely. :)
I'm glad you brought this up! I'll try to follow up with a PR to make this part of [datafusion-table-providers](https://github.com/datafusion-contrib/datafusion-table-providers), where it most likely belongs.
As a side note, I only published it here initially because I noticed that https://github.com/apache/datafusion/issues/7731 was open and acknowledged.
